lib/mlir/Dialect/Clift/IR/CliftTypes.cpp

```diff
@@ -2,58 +2,728 @@
+
+bool TypedefType::getAlias(llvm::raw_ostream &OS) const {
+  if (getIsConst())
+    return false;
+
+  return getDefinedTypeAlias(*this, OS);
+}
+
+uint64_t TypedefType::getByteSize() const {
+  return getUnderlyingType().getByteSize();
+}
+
+clift::ValueType TypedefType::addConst() const {
+  return Base::get(getContext(), getDefinition(), /*IsConst=*/true);
+}
+
+clift::ValueType TypedefType::removeConst() const {
+  return Base::get(getContext(), getDefinition(), /*IsConst=*/false);
+}
+
+mlir::Type TypedefType::parse(mlir::AsmParser &Parser) {
+  if (Parser.parseLess().failed())
+    return {};
+
+  std::string Handle;
+  if (Parser.parseString(&Handle).failed())
+    return {};
+
+  std::string Name;
+  if (Parser.parseOptionalKeyword("as").succeeded()) {
+    if (Parser.parseString(&Name).failed())
+      return {};
+  }
+
+  if (Parser.parseColon().failed())
+    return {};
+
+  clift::ValueType UnderlyingType;
+  if (Parser.parseType(UnderlyingType).failed())
+    return {};
+
+  if (Parser.parseGreater().failed())
+    return {};
+
+  return TypedefType::get(Parser.getContext(), Handle, Name, UnderlyingType);
+}
+
+void TypedefType::print(mlir::AsmPrinter &Printer) const {
+  Printer << "<";
+  printString(Printer, getHandle());
+  printDefinedTypeName(Printer, getName());
+
+  Printer << " : ";
+  Printer.printType(getUnderlyingType());
+  Printer << ">";
+}
+
+//===---------------------------- FunctionType ----------------------------===//
+
+mlir::LogicalResult FunctionType::verify(EmitErrorType EmitError,
+                                         llvm::StringRef Handle,
+                                         llvm::StringRef Name,
+                                         mlir::Type ReturnType,
+                                         llvm::ArrayRef<mlir::Type> Args) {
+  auto R = mlir::dyn_cast<clift::ValueType>(ReturnType);
+  if (not R)
+    return EmitError() << "Function return type must be a ValueType";
+
+  for (mlir::Type ArgumentType : Args) {
+    auto T = mlir::dyn_cast<clift::ValueType>(ArgumentType);
+    if (not T)
+      return EmitError() << "Function argument types must be ValueTypes";
+
+    if (not isObjectType(T))
+      return EmitError() << "Function parameter type must be an object type";
+    if (isArrayType(T))
+      return EmitError() << "Function parameter type may not be an array type";
+  }
+
+  if (not isReturnableType(R))
+    return EmitError() << "Function return type must be void or a non-array "
+                          "object type.";
+
+  return mlir::success();
+}
+
+mlir::LogicalResult
+FunctionType::verify(EmitErrorType EmitError,
+                     llvm::StringRef Handle,
+                     llvm::StringRef Name,
+                     clift::ValueType ReturnType,
+                     llvm::ArrayRef<clift::ValueType> Args) {
+  if (not isReturnableType(ReturnType))
+    return EmitError() << "Function return type must be void or a non-array "
```

Comment:
If you change `isReturnableType` you should fix this error message too.

Reply:
Also, perhaps besides `isReturnableType` you should have a `verifyReturnableType`, returning a `mlir::LogicalResult` directly, so you can have the error string in only one place.